File: packages/web/TraverseEmbedder/src/hostConnectorCommand.ts
/**
 * Shared Spec 137 host-connector command/event contract for browser and macOS.
 * Host adapters stay out of this module; they remain platform implementations.
 */

export const HOST_CONNECTOR_COMMAND_SCHEMA_VERSION = "1.0.0" as const;
export const HOST_CONNECTOR_COMMAND_KIND = "host_connector_command" as const;
export const HOST_CONNECTOR_RESULT_KIND = "host_connector_result" as const;
export const HOST_CONNECTOR_EVENT_KIND = "host_connector_event" as const;
export const HOST_CONNECTOR_GOVERNING_SPEC =
  "137-host-connector-command-dispatch" as const;
export const AUDIO_INPUT_CONNECTOR = "traverse.audio-input" as const;
export const AUDIO_CAPTURE_OPERATION = "audio.capture" as const;
export const AUDIO_PERMISSION_REQUEST_OPERATION =
  "audio.permission.request" as const;
export const MODEL_RUNTIME_CONNECTOR = "traverse.model-runtime" as const;
export const MODEL_EXECUTE_OPERATION = "model.execute" as const;

export type HostConnectorTargetFamily = "browser" | "macos" | "local";

export type HostConnectorErrorCode =
  | "unknown_command"
  | "unbound"
  | "incompatible"
  | "unconfigured"
  | "target_incompatible"
  | "input_limit_exceeded"
  | "cancelled"
  | "idempotency_conflict"
  | "policy_denied"
  | "unavailable"
  | "invalid_input"
  | "model_unavailable"
  | "model_incompatible"
  | "resource_exhausted"
  | "timeout"
  | "execution_failed";

/** Spec 138 exact model pin carried on `model.execute`. */
export interface ModelRef {
  readonly model_id: string;
  readonly version: string;
  readonly digest: string;
}

/** Spec 138 `model.execute` request payload (tensors via host-staged refs). */
export interface ModelExecutePayload {
  readonly model_ref: ModelRef;
  readonly input_ref: string;
  readonly policy_ref: string;
  readonly data_classification: string;
  readonly input_schema_ref: string;
  readonly input_schema_version: string;
  readonly max_output_bytes: number;
  readonly max_memory_bytes?: number;
  readonly max_fuel?: number;
  readonly timeout_ms?: number;
  readonly feature_metadata?: Record<string, unknown>;
}

export const MODEL_RUNTIME_GOVERNING_SPEC =
  "138-governed-exact-model-execution" as const;
export const PLACEMENT_WASM_CPU = "wasm-cpu" as const;

/**
 * Normalize a Spec 138 model.execute success/failure for cross-target compare.
 * Strips host-private fields; keeps status-bearing public codes and identity.
 */
export function normalizeModelExecuteEvidence(input: {
  readonly status?: string;
  readonly error_code?: HostConnectorErrorCode;
  readonly model_ref?: ModelRef;
  readonly placement?: string;
  readonly output_ref?: string;
  readonly artifact_ref?: string;
}): Record<string, unknown> {
  return {
    governing_spec: MODEL_RUNTIME_GOVERNING_SPEC,
    status: input.status ?? (input.error_code ? "failed" : "ok"),
    error_code: input.error_code ?? null,
    model_ref: input.model_ref ?? null,
    placement: input.placement ?? PLACEMENT_WASM_CPU,
    has_output_ref: Boolean(input.output_ref ?? input.artifact_ref),
  };
}

export function modelExecuteCommand(
  commandId: string,
  correlationId: string,
  idempotencyKey: string,
  targetFamily: HostConnectorTargetFamily,
  payload: ModelExecutePayload,
): HostConnectorAppCommand {
  return {
    kind: HOST_CONNECTOR_COMMAND_KIND,
    schema_version: HOST_CONNECTOR_COMMAND_SCHEMA_VERSION,
    command: "run_local_model",
    command_id: commandId,
    correlation_id: correlationId,
    idempotency_key: idempotencyKey,
    target_family: targetFamily,
    cancel_requested: false,
    payload: { ...payload },
  };
}

export type HostConnectorEventName =
  | "accepted"
  | "started"
  | "completed"
  | "cancelled"
  | "failed";

export interface HostConnectorAppCommand {
  readonly kind: typeof HOST_CONNECTOR_COMMAND_KIND;
  readonly schema_version: typeof HOST_CONNECTOR_COMMAND_SCHEMA_VERSION;
  readonly command: string;
  readonly command_id: string;
  readonly correlation_id: string;
  readonly idempotency_key: string;
  readonly target_family: HostConnectorTargetFamily;
  readonly cancel_requested?: boolean;
  readonly payload: Record<string, unknown>;
}

export interface HostConnectorEvent {
  readonly kind: typeof HOST_CONNECTOR_EVENT_KIND;
  readonly schema_version: typeof HOST_CONNECTOR_COMMAND_SCHEMA_VERSION;
  readonly event: HostConnectorEventName;
  readonly command_id: string;
  readonly correlation_id: string;
  readonly connector_id?: string;
  readonly operation?: string;
  readonly binding_id?: string;
  readonly target_family: HostConnectorTargetFamily;
  readonly artifact_ref?: string;
  readonly error_code?: HostConnectorErrorCode;
}

export interface HostConnectorError {
  readonly code: HostConnectorErrorCode;
  readonly message: string;
}

export function audioCaptureCommand(
  commandId: string,
  correlationId: string,
  idempotencyKey: string,
  targetFamily: HostConnectorTargetFamily,
  payload: { readonly max_duration_ms: number; readonly max_bytes: number },
): HostConnectorAppCommand {
  return {
    kind: HOST_CONNECTOR_COMMAND_KIND,
    schema_version: HOST_CONNECTOR_COMMAND_SCHEMA_VERSION,
    command: "capture_audio",
    command_id: commandId,
    correlation_id: correlationId,
    idempotency_key: idempotencyKey,
    target_family: targetFamily,
    cancel_requested: false,
    payload: { ...payload },
  };
}

export function audioPermissionCommand(
  commandId: string,
  correlationId: string,
  idempotencyKey: string,
  targetFamily: HostConnectorTargetFamily,
): HostConnectorAppCommand {
  return {
    kind: HOST_CONNECTOR_COMMAND_KIND,
    schema_version: HOST_CONNECTOR_COMMAND_SCHEMA_VERSION,
    command: "request_permission",
    command_id: commandId,
    correlation_id: correlationId,
    idempotency_key: idempotencyKey,
    target_family: targetFamily,
    cancel_requested: false,
    payload: {},
  };
}

File: packages/swift/TraverseEmbedder/Sources/TraverseEmbedder/HostConnectorCommand.swift
import Foundation

/// Shared Spec 137 host-connector command/event contract for browser and macOS.
/// Native/browser adapters remain host implementations behind this wire type.

public enum HostConnectorCommandContract {
    public static let schemaVersion = "1.0.0"
    public static let commandKind = "host_connector_command"
    public static let resultKind = "host_connector_result"
    public static let eventKind = "host_connector_event"
    public static let governingSpec = "137-host-connector-command-dispatch"
    public static let audioInputConnector = "traverse.audio-input"
    public static let audioCaptureOperation = "audio.capture"
    public static let modelRuntimeConnector = "traverse.model-runtime"
    public static let modelExecuteOperation = "model.execute"
}

public struct HostConnectorAppCommand: Sendable, Equatable {
    public let kind: String
    public let schemaVersion: String
    public let command: String
    public let commandId: String
    public let correlationId: String
    public let idempotencyKey: String
    public let targetFamily: String
    public let cancelRequested: Bool
    public let payloadJSON: Data

    public init(
        command: String,
        commandId: String,
        correlationId: String,
        idempotencyKey: String,
        targetFamily: String,
        cancelRequested: Bool = false,
        payloadJSON: Data
    ) {
        self.kind = HostConnectorCommandContract.commandKind
        self.schemaVersion = HostConnectorCommandContract.schemaVersion
        self.command = command
        self.commandId = commandId
        self.correlationId = correlationId
        self.idempotencyKey = idempotencyKey
        self.targetFamily = targetFamily
        self.cancelRequested = cancelRequested
        self.payloadJSON = payloadJSON
    }

    public static func audioCapture(
        commandId: String,
        correlationId: String,
        idempotencyKey: String,
        targetFamily: String,
        maxDurationMs: Int,
        maxBytes: Int
    ) throws -> HostConnectorAppCommand {
        let payload = try JSONSerialization.data(
            withJSONObject: ["max_duration_ms": maxDurationMs, "max_bytes": maxBytes]
        )
        return HostConnectorAppCommand(
            command: "capture_audio",
            commandId: commandId,
            correlationId: correlationId,
            idempotencyKey: idempotencyKey,
            targetFamily: targetFamily,
            payloadJSON: payload
        )
    }

    /// Snake_case wire object shared with the browser embedder.
    public func wireObject() throws -> [String: Any] {
        let payload = try JSONSerialization.jsonObject(with: payloadJSON)
        return [
            "kind": kind,
            "schema_version": schemaVersion,
            "command": command,
            "command_id": commandId,
            "correlation_id": correlationId,
            "idempotency_key": idempotencyKey,
            "target_family": targetFamily,
            "cancel_requested": cancelRequested,
            "payload": payload,
        ]
    }
}

public struct HostConnectorEvent: Sendable, Equatable {
    public let kind: String
    public let schemaVersion: String
    public let event: String
    public let commandId: String
    public let correlationId: String
    public let connectorId: String?
    public let operation: String?
    public let bindingId: String?
    public let targetFamily: String
    public let artifactRef: String?
    public let errorCode: String?

    public init(
        event: String,
        commandId: String,
        correlationId: String,
        targetFamily: String,
        connectorId: String? = nil,
        operation: String? = nil,
        bindingId: String? = nil,
        artifactRef: String? = nil,
        errorCode: String? = nil
    ) {
        self.kind = HostConnectorCommandContract.eventKind
        self.schemaVersion = HostConnectorCommandContract.schemaVersion
        self.event = event
        self.commandId = commandId
        self.correlationId = correlationId
        self.connectorId = connectorId
        self.operation = operation
        self.bindingId = bindingId
        self.targetFamily = targetFamily
        self.artifactRef = artifactRef
        self.errorCode = errorCode
    }

    /// Snake_case wire object shared with the browser embedder.
    public func wireObject() -> [String: Any] {
        var object: [String: Any] = [
            "kind": kind,
            "schema_version": schemaVersion,
            "event": event,
            "command_id": commandId,
            "correlation_id": correlationId,
            "target_family": targetFamily,
        ]
        if let connectorId { object["connector_id"] = connectorId }
        if let operation { object["operation"] = operation }
        if let bindingId { object["binding_id"] = bindingId }
        if let artifactRef { object["artifact_ref"] = artifactRef }
        if let errorCode { object["error_code"] = errorCode }
        return object
    }
}
